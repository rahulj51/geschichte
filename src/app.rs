use crate::cache::DiffCache;
use crate::commit::Commit;
use crate::error::Result;
use std::collections::HashMap;
use std::path::PathBuf;

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum FocusedPanel {
    Commits,
    Diff,
}

pub struct App {
    pub repo_root: PathBuf,
    pub file_path: PathBuf,
    pub should_quit: bool,
    pub context_lines: u32,
    pub follow_renames: bool,
    pub first_parent: bool,
    
    // Commit data
    pub commits: Vec<Commit>,
    pub selected_index: usize,
    pub rename_map: HashMap<String, PathBuf>,
    
    // Diff display
    pub current_diff: String,
    pub diff_scroll: usize,
    pub diff_cache: DiffCache,
    
    // UI state
    pub focused_panel: FocusedPanel,
    pub split_ratio: f32,
    pub show_help: bool,
    pub loading: bool,
    pub error_message: Option<String>,
}

impl App {
    pub fn new(
        repo_root: PathBuf,
        file_path: PathBuf,
        context_lines: u32,
        follow_renames: bool,
        first_parent: bool,
    ) -> Self {
        Self {
            repo_root,
            file_path,
            should_quit: false,
            context_lines,
            follow_renames,
            first_parent,
            commits: Vec::new(),
            selected_index: 0,
            rename_map: HashMap::new(),
            current_diff: String::new(),
            diff_scroll: 0,
            diff_cache: DiffCache::new(50),
            focused_panel: FocusedPanel::Commits,
            split_ratio: 0.4, // 40% commits, 60% diff
            show_help: false,
            loading: false,
            error_message: None,
        }
    }

    pub fn load_git_data(&mut self) -> Result<()> {
        self.loading = true;
        self.error_message = None;

        // Load commits
        let mut commits = crate::git::history::fetch_commit_history(
            &self.repo_root,
            &self.file_path,
            self.follow_renames,
            self.first_parent,
        )?;

        // Check for working directory changes and prepend if found
        let wd_status = crate::git::working::check_working_directory_status(
            &self.repo_root,
            &self.file_path,
        )?;

        if wd_status != crate::git::working::WorkingDirectoryStatus::Clean {
            let status_text = match wd_status {
                crate::git::working::WorkingDirectoryStatus::Modified => "Modified".to_string(),
                crate::git::working::WorkingDirectoryStatus::Staged => "Staged".to_string(),
                crate::git::working::WorkingDirectoryStatus::ModifiedAndStaged => "Modified + Staged".to_string(),
                crate::git::working::WorkingDirectoryStatus::Clean => unreachable!(),
            };
            
            let wd_commit = crate::commit::Commit::new_working_directory(status_text);
            commits.insert(0, wd_commit);
        }

        self.commits = commits;

        // Build rename map
        if self.follow_renames {
            self.rename_map = crate::git::history::build_rename_map(
                &self.repo_root,
                &self.file_path,
            )?;
        }

        // Load initial diff if we have commits
        if !self.commits.is_empty() {
            self.load_diff_for_selected_commit()?;
        }

        self.loading = false;
        Ok(())
    }

    pub fn load_diff_for_selected_commit(&mut self) -> Result<()> {
        if self.commits.is_empty() || self.selected_index >= self.commits.len() {
            return Ok(());
        }

        let commit = &self.commits[self.selected_index];
        
        // Check cache first
        if let Some(cached_diff) = self.diff_cache.get(&commit.hash).cloned() {
            self.current_diff = cached_diff;
            return Ok(());
        }

        let diff = if commit.is_working_directory {
            // Handle working directory diff
            crate::git::working::fetch_working_directory_diff(
                &self.repo_root,
                &self.file_path,
                self.context_lines,
            )?
        } else {
            // Handle regular commit diff
            let parents = crate::git::history::get_commit_parents(&self.repo_root, &commit.hash)?;
            let parent_hash = parents.first().map(|s| s.as_str());

            // Resolve file path at this commit
            let file_path = self.rename_map
                .get(&commit.hash)
                .cloned()
                .unwrap_or_else(|| self.file_path.clone());

            crate::git::diff::fetch_diff(
                &self.repo_root,
                &commit.hash,
                parent_hash,
                &file_path,
                self.context_lines,
            )?
        };

        // Cache and store
        self.diff_cache.put(commit.hash.clone(), diff.clone());
        self.current_diff = diff;

        Ok(())
    }

    pub fn quit(&mut self) {
        self.should_quit = true;
    }

    pub fn move_selection_up(&mut self) -> Result<()> {
        if self.selected_index > 0 {
            self.selected_index -= 1;
            self.load_diff_for_selected_commit()?;
        }
        Ok(())
    }

    pub fn move_selection_down(&mut self) -> Result<()> {
        if self.selected_index + 1 < self.commits.len() {
            self.selected_index += 1;
            self.load_diff_for_selected_commit()?;
        }
        Ok(())
    }

    pub fn scroll_diff_up(&mut self) {
        if self.diff_scroll > 0 {
            self.diff_scroll -= 1;
        }
    }

    pub fn scroll_diff_down(&mut self) {
        self.diff_scroll += 1;
    }

    pub fn switch_focus(&mut self) {
        self.focused_panel = match self.focused_panel {
            FocusedPanel::Commits => FocusedPanel::Diff,
            FocusedPanel::Diff => FocusedPanel::Commits,
        };
    }

    pub fn increase_split_ratio(&mut self) {
        self.split_ratio = (self.split_ratio + 0.05).min(0.7);
    }

    pub fn decrease_split_ratio(&mut self) {
        self.split_ratio = (self.split_ratio - 0.05).max(0.2);
    }

    pub fn toggle_help(&mut self) {
        self.show_help = !self.show_help;
    }

    pub fn handle_key(&mut self, key: crossterm::event::KeyEvent) -> Result<()> {
        use crossterm::event::{KeyCode, KeyModifiers};

        match (key.code, key.modifiers) {
            (KeyCode::Char('q'), KeyModifiers::NONE) => self.quit(),
            (KeyCode::Esc, _) => {
                if self.show_help {
                    self.show_help = false;
                } else {
                    self.quit();
                }
            }
            (KeyCode::Tab, KeyModifiers::NONE) => {
                self.switch_focus();
            }
            (KeyCode::Up, KeyModifiers::NONE) | (KeyCode::Char('k'), KeyModifiers::NONE) => {
                match self.focused_panel {
                    FocusedPanel::Commits => self.move_selection_up()?,
                    FocusedPanel::Diff => self.scroll_diff_up(),
                }
            }
            (KeyCode::Down, KeyModifiers::NONE) | (KeyCode::Char('j'), KeyModifiers::NONE) => {
                match self.focused_panel {
                    FocusedPanel::Commits => self.move_selection_down()?,
                    FocusedPanel::Diff => self.scroll_diff_down(),
                }
            }
            (KeyCode::PageUp, _) => {
                // Always scroll diff for PageUp/PageDown regardless of focus
                for _ in 0..5 {
                    self.scroll_diff_up();
                }
            }
            (KeyCode::PageDown, _) => {
                // Always scroll diff for PageUp/PageDown regardless of focus
                for _ in 0..5 {
                    self.scroll_diff_down();
                }
            }
            // Mac-friendly vim-style navigation
            (KeyCode::Char('u'), KeyModifiers::CONTROL) => {
                // Ctrl+U = Page Up (vim-style)
                for _ in 0..5 {
                    self.scroll_diff_up();
                }
            }
            (KeyCode::Char('d'), KeyModifiers::CONTROL) => {
                // Ctrl+D = Page Down (vim-style)
                for _ in 0..5 {
                    self.scroll_diff_down();
                }
            }
            // Mac-friendly emacs-style navigation
            (KeyCode::Char('b'), KeyModifiers::CONTROL) => {
                // Ctrl+B = Page Up (emacs-style)
                for _ in 0..5 {
                    self.scroll_diff_up();
                }
            }
            (KeyCode::Char('f'), KeyModifiers::CONTROL) => {
                // Ctrl+F = Page Down (emacs-style)
                for _ in 0..5 {
                    self.scroll_diff_down();
                }
            }
            (KeyCode::Char('h'), KeyModifiers::NONE) => {
                self.decrease_split_ratio();
            }
            (KeyCode::Char('l'), KeyModifiers::NONE) => {
                self.increase_split_ratio();
            }
            (KeyCode::Char('/'), KeyModifiers::NONE) => {
                // TODO: Start search
            }
            (KeyCode::Char('?'), KeyModifiers::NONE) => {
                self.toggle_help();
            }
            _ => {}
        }

        Ok(())
    }
}