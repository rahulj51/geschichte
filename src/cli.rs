use clap::{Parser, ValueEnum};
use std::path::PathBuf;

#[derive(Clone, Copy, Debug, PartialEq, ValueEnum)]
pub enum LayoutMode {
    /// Traditional unified diff view (two panels)
    Unified,
    /// Side-by-side diff view (three panels)
    SideBySide,
    /// Automatically choose based on terminal size
    Auto,
}

#[derive(Parser, Debug)]
#[command(
    name = "geschichte",
    version,
    about = "A blazingly fast TUI for viewing git file history and diffs",
    long_about = None
)]
pub struct Args {
    /// Path to the file to view history for (optional - opens file picker if not provided)
    #[arg(value_name = "FILE")]
    pub file_path: Option<PathBuf>,

    /// Repository root directory (auto-discovered if not specified)
    #[arg(short = 'C', long = "repo", value_name = "DIR")]
    pub repo_path: Option<PathBuf>,

    /// Number of context lines in diffs
    #[arg(short = 'L', long = "lines", default_value = "3")]
    pub context_lines: u32,

    /// Show full file content in diffs instead of just context around changes
    #[arg(long = "full-file")]
    pub full_file: bool,

    /// Show only first-parent commits (linearize merge commits)
    #[arg(long = "first-parent")]
    pub first_parent: bool,

    /// Disable rename tracking
    #[arg(long = "no-follow")]
    pub no_follow: bool,

    /// Enable debug logging
    #[arg(long = "debug")]
    pub debug: bool,

    /// Enable side-by-side diff view (three-panel layout)
    #[arg(short = 's', long = "side-by-side")]
    pub side_by_side: bool,

    /// Layout mode for the UI
    #[arg(long = "layout", value_enum, default_value = "unified")]
    pub layout: LayoutMode,
}

impl Args {
    pub fn validate(&self) -> Result<(), String> {
        if !self.full_file && self.context_lines > 100 {
            return Err("Context lines must be between 0 and 100".to_string());
        }

        Ok(())
    }

    /// Get the effective context lines, considering the full-file flag
    pub fn effective_context_lines(&self) -> u32 {
        if self.full_file {
            // Use a very large number to show the full file
            9999
        } else {
            self.context_lines
        }
    }

    /// Get the effective layout mode, considering both --side-by-side flag and --layout option
    pub fn effective_layout(&self) -> LayoutMode {
        // --side-by-side flag takes precedence for backwards compatibility
        if self.side_by_side {
            LayoutMode::SideBySide
        } else {
            self.layout
        }
    }
}
