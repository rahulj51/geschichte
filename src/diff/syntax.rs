use once_cell::sync::Lazy;
use ratatui::style::{Color as RatatuiColor, Modifier};
use ratatui::text::Span;
use std::path::Path;
use syntect::easy::HighlightLines;
use syntect::highlighting::ThemeSet;
use syntect::parsing::{SyntaxReference, SyntaxSet};

/// Lazy-loaded syntax definitions
static SYNTAX_SET: Lazy<SyntaxSet> = Lazy::new(SyntaxSet::load_defaults_newlines);

/// Lazy-loaded themes
static THEME_SET: Lazy<ThemeSet> = Lazy::new(ThemeSet::load_defaults);

/// Highlight a line of code for a given file path
pub fn highlight_line(line: &str, file_path: &Path) -> Vec<Span<'static>> {
    let syntax = detect_syntax(file_path);

    if let Some(syntax) = syntax {
        let theme = &THEME_SET.themes["InspiredGitHub"]; // Try a light theme designed for GitHub
        let mut highlighter = HighlightLines::new(syntax, theme);

        match highlighter.highlight_line(line, &SYNTAX_SET) {
            Ok(ranges) => {
                ranges
                    .into_iter()
                    .map(|(style, text)| {
                        let color = syntect_to_ratatui_color(style.foreground);
                        let mut ratatui_style = ratatui::style::Style::default().fg(color);

                        // Add modifiers based on syntect style
                        if style
                            .font_style
                            .contains(syntect::highlighting::FontStyle::BOLD)
                        {
                            ratatui_style = ratatui_style.add_modifier(Modifier::BOLD);
                        }
                        if style
                            .font_style
                            .contains(syntect::highlighting::FontStyle::ITALIC)
                        {
                            ratatui_style = ratatui_style.add_modifier(Modifier::ITALIC);
                        }
                        if style
                            .font_style
                            .contains(syntect::highlighting::FontStyle::UNDERLINE)
                        {
                            ratatui_style = ratatui_style.add_modifier(Modifier::UNDERLINED);
                        }

                        Span::styled(text.to_string(), ratatui_style)
                    })
                    .collect()
            }
            Err(_) => {
                // Fallback to plain text
                vec![Span::raw(line.to_string())]
            }
        }
    } else {
        // No syntax highlighting available
        vec![Span::raw(line.to_string())]
    }
}

/// Detect syntax from file extension
fn detect_syntax(file_path: &Path) -> Option<&'static SyntaxReference> {
    let extension = file_path.extension()?.to_str()?;

    // Try to find syntax by extension
    SYNTAX_SET
        .find_syntax_by_extension(extension)
        .or_else(|| {
            // Fallback to file name
            let file_name = file_path.file_name()?.to_str()?;
            SYNTAX_SET.find_syntax_by_name(file_name)
        })
        .or_else(|| {
            // Special cases
            match extension {
                "rs" => SYNTAX_SET.find_syntax_by_name("Rust"),
                "py" => SYNTAX_SET.find_syntax_by_name("Python"),
                "js" | "jsx" => SYNTAX_SET.find_syntax_by_name("JavaScript"),
                "ts" | "tsx" => SYNTAX_SET.find_syntax_by_name("TypeScript"),
                "md" => SYNTAX_SET.find_syntax_by_name("Markdown"),
                "toml" => SYNTAX_SET.find_syntax_by_name("TOML"),
                "yaml" | "yml" => SYNTAX_SET.find_syntax_by_name("YAML"),
                "json" => SYNTAX_SET.find_syntax_by_name("JSON"),
                "sh" | "bash" => SYNTAX_SET.find_syntax_by_name("Bash"),
                "go" => SYNTAX_SET.find_syntax_by_name("Go"),
                "cpp" | "cc" | "cxx" => SYNTAX_SET.find_syntax_by_name("C++"),
                "c" | "h" => SYNTAX_SET.find_syntax_by_name("C"),
                _ => None,
            }
        })
}

/// Convert syntect color to ratatui color
fn syntect_to_ratatui_color(color: syntect::highlighting::Color) -> RatatuiColor {
    RatatuiColor::Rgb(color.r, color.g, color.b)
}
