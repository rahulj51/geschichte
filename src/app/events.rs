use crate::app::{App, AppMode, FilePickerContext, FocusedPanel};
use crate::error::Result;
use crossterm::event::{KeyCode, KeyEvent, KeyModifiers};

impl App {
    pub fn handle_navigation_keys(&mut self, key: KeyEvent) -> Result<bool> {
        match (key.code, key.modifiers) {
            (KeyCode::Up, KeyModifiers::NONE) | (KeyCode::Char('k'), KeyModifiers::NONE) => {
                if self.show_commit_info {
                    self.scroll_commit_info_up();
                } else if let Some(focused_panel) = self.get_focused_panel() {
                    match focused_panel {
                        FocusedPanel::Commits => self.move_selection_up()?,
                        FocusedPanel::Diff => {
                            let layout_mode = self.effective_layout();
                            self.ui_state.move_cursor_up(&layout_mode);
                        }
                    }
                }
                Ok(true)
            }
            (KeyCode::Down, KeyModifiers::NONE) | (KeyCode::Char('j'), KeyModifiers::NONE) => {
                if self.show_commit_info {
                    self.scroll_commit_info_down();
                } else if let Some(focused_panel) = self.get_focused_panel() {
                    match focused_panel {
                        FocusedPanel::Commits => self.move_selection_down()?,
                        FocusedPanel::Diff => {
                            let max_lines = self.get_diff_line_count();
                            let layout_mode = self.effective_layout();
                            self.ui_state.move_cursor_down(max_lines, &layout_mode);
                        }
                    }
                }
                Ok(true)
            }
            (KeyCode::Tab, KeyModifiers::NONE) => {
                self.switch_focus();
                Ok(true)
            }
            _ => Ok(false),
        }
    }

    pub fn handle_scrolling_keys(&mut self, key: KeyEvent) -> Result<bool> {
        match (key.code, key.modifiers) {
            (KeyCode::PageUp, _) => {
                // Always scroll diff for PageUp/PageDown regardless of focus
                self.ui_state.scroll_diff_page_up();
                Ok(true)
            }
            (KeyCode::PageDown, _) => {
                // Always scroll diff for PageUp/PageDown regardless of focus
                let max_lines = self.get_diff_line_count();
                self.ui_state.scroll_diff_page_down(max_lines);
                Ok(true)
            }
            // Mac-friendly vim-style navigation
            (KeyCode::Char('u'), KeyModifiers::CONTROL) => {
                // Ctrl+U = Page Up (vim-style)
                self.ui_state.scroll_diff_page_up();
                Ok(true)
            }
            (KeyCode::Char('d'), KeyModifiers::CONTROL) => {
                // Ctrl+D = Page Down (vim-style)
                let max_lines = self.get_diff_line_count();
                self.ui_state.scroll_diff_page_down(max_lines);
                Ok(true)
            }
            // Mac-friendly emacs-style navigation
            (KeyCode::Char('b'), KeyModifiers::CONTROL) => {
                // Ctrl+B = Page Up (emacs-style)
                self.ui_state.scroll_diff_page_up();
                Ok(true)
            }
            (KeyCode::Char('f'), KeyModifiers::CONTROL) => {
                // Ctrl+F = Page Down (emacs-style)
                let max_lines = self.get_diff_line_count();
                self.ui_state.scroll_diff_page_down(max_lines);
                Ok(true)
            }
            // Horizontal scrolling (but not when in copy mode)
            (KeyCode::Char('a'), KeyModifiers::NONE) => {
                // Don't handle 'a' for scrolling when in copy mode
                if self.copy_mode.is_some() {
                    return Ok(false); // Let copy handler deal with it
                }
                if let Some(focused_panel) = self.get_focused_panel() {
                    match focused_panel {
                        FocusedPanel::Commits => self.ui_state.scroll_commit_left(),
                        FocusedPanel::Diff => self.ui_state.scroll_diff_left(),
                    }
                }
                Ok(true)
            }
            (KeyCode::Char('s'), KeyModifiers::NONE) => {
                // Don't handle 's' for scrolling when in copy mode
                if self.copy_mode.is_some() {
                    return Ok(false); // Let copy handler deal with it
                }
                if let Some(focused_panel) = self.get_focused_panel() {
                    match focused_panel {
                        FocusedPanel::Commits => {
                            let max_width = self.calculate_max_commit_line_width();
                            self.ui_state.scroll_commit_right(max_width);
                        }
                        FocusedPanel::Diff => {
                            let max_width = self.calculate_max_diff_line_width();
                            self.ui_state.scroll_diff_right(max_width);
                        }
                    }
                }
                Ok(true)
            }
            _ => Ok(false),
        }
    }

    pub fn handle_ui_keys(&mut self, key: KeyEvent) -> Result<bool> {
        match (key.code, key.modifiers) {
            (KeyCode::Char('q'), KeyModifiers::NONE) => {
                if self.show_commit_info {
                    self.hide_commit_info_popup();
                } else {
                    self.quit();
                }
                Ok(true)
            }
            (KeyCode::Esc, _) => {
                if self.ui_state.show_help {
                    self.ui_state.show_help = false;
                } else if self.show_commit_info {
                    self.hide_commit_info_popup();
                } else if self.diff_search_state.is_some() {
                    self.clear_diff_search();
                } else if self.copy_mode.is_some() {
                    self.cancel_copy_mode();
                } else if self.diff_range_start.is_some() {
                    self.clear_diff_range_selection();
                } else {
                    // HACK: revert the FilePicker context to Initial.
                    use crate::app::FilePickerState;
                    use crate::git::files::get_git_files;
                    let files = get_git_files(&self.repo_root)?;
                    self.mode = AppMode::FilePicker {
                        state: FilePickerState::new(files),
                        context: FilePickerContext::Initial,
                    };
                    // FIX: I wish it could be more like this tho.
                    // self.mode = AppMode::FilePicker {
                    //     state: mode.state,
                    //     context: FilePickerContext::Initial,
                    // };
                    if let Err(e) = self.switch_to_file_picker() {
                        self.error_message = Some(format!("Failed to open file picker: {}", e));
                    }
                }
                // else {
                //     self.quit();
                // }
                Ok(true)
            }
            (KeyCode::Char('h'), KeyModifiers::NONE) => {
                self.ui_state.decrease_split_ratio();
                Ok(true)
            }
            (KeyCode::Char('l'), KeyModifiers::NONE) => {
                self.ui_state.increase_split_ratio();
                Ok(true)
            }
            (KeyCode::Char('f'), KeyModifiers::NONE) => {
                // Open file picker to switch files
                if let Err(e) = self.switch_to_file_picker() {
                    self.error_message = Some(format!("Failed to open file picker: {}", e));
                }
                Ok(true)
            }
            (KeyCode::Char('/'), KeyModifiers::NONE) => {
                if !self.show_commit_info && self.get_focused_panel() == Some(FocusedPanel::Diff) {
                    self.start_diff_search();
                    Ok(true)
                } else {
                    Ok(false) // Let other handlers deal with it
                }
            }
            (KeyCode::Char('d'), KeyModifiers::NONE) => {
                self.toggle_diff_range_selection()?;
                Ok(true)
            }
            (KeyCode::Char('?'), KeyModifiers::NONE) => {
                self.ui_state.toggle_help();
                Ok(true)
            }
            (KeyCode::Char('i'), KeyModifiers::NONE) | (KeyCode::Enter, KeyModifiers::NONE) => {
                // Show commit info popup (only in commits panel)
                if matches!(self.get_focused_panel(), Some(FocusedPanel::Commits)) {
                    self.show_commit_info_popup()?;
                }
                Ok(true)
            }
            _ => Ok(false),
        }
    }

    pub fn handle_copy_keys(&mut self, key: KeyEvent) -> Result<bool> {
        // Handle copy keys in commits panel and history mode, or in commit info popup
        if !matches!(self.get_focused_panel(), Some(FocusedPanel::Commits))
            && !self.show_commit_info
        {
            return Ok(false);
        }

        match (key.code, key.modifiers) {
            (KeyCode::Char('y'), KeyModifiers::NONE) => {
                match self.copy_mode.as_ref() {
                    None => {
                        // First 'y' press - start copy mode
                        self.start_copy_mode();
                    }
                    Some(crate::copy::CopyMode::WaitingForTarget) => {
                        // Second 'y' press - copy full SHA
                        self.copy_commit_sha(false)?;
                    }
                }
                Ok(true)
            }
            (KeyCode::Char('Y'), KeyModifiers::SHIFT) => {
                // Capital Y - copy short SHA directly
                self.copy_commit_sha(true)?;
                Ok(true)
            }
            (KeyCode::Char('c'), KeyModifiers::NONE) => {
                // Direct copy of full SHA (especially useful in popup)
                if self.show_commit_info {
                    self.copy_commit_sha(false)?;
                } else {
                    // Start copy mode in normal view
                    self.start_copy_mode();
                }
                Ok(true)
            }
            // Note: 'm' key is only handled in copy mode section below
            _ => {
                // Handle copy mode targets
                if matches!(
                    self.copy_mode,
                    Some(crate::copy::CopyMode::WaitingForTarget)
                ) {
                    match (key.code, key.modifiers) {
                        (KeyCode::Char('s'), KeyModifiers::NONE) => {
                            self.copy_commit_sha(false)?;
                            Ok(true)
                        }
                        (KeyCode::Char('h'), KeyModifiers::NONE) => {
                            self.copy_commit_sha(true)?;
                            Ok(true)
                        }
                        (KeyCode::Char('m'), KeyModifiers::NONE) => {
                            self.copy_commit_message()?;
                            Ok(true)
                        }
                        (KeyCode::Char('a'), KeyModifiers::NONE) => {
                            self.copy_commit_author()?;
                            Ok(true)
                        }
                        (KeyCode::Char('d'), KeyModifiers::NONE) => {
                            self.copy_commit_date()?;
                            Ok(true)
                        }
                        (KeyCode::Char('u'), KeyModifiers::NONE) => {
                            self.copy_github_url()?;
                            Ok(true)
                        }
                        (KeyCode::Char('p'), KeyModifiers::NONE) => {
                            self.copy_file_relative_path()?;
                            Ok(true)
                        }

                        _ => Ok(false),
                    }
                } else {
                    Ok(false)
                }
            }
        }
    }

    pub fn handle_change_navigation_keys(&mut self, key: KeyEvent) -> Result<bool> {
        // Check if we're in active search mode first
        if let Some(ref search_state) = self.diff_search_state {
            if search_state.is_active
                && !search_state.is_input_mode
                && !search_state.results.is_empty()
            {
                match (key.code, key.modifiers) {
                    (KeyCode::Char('n'), KeyModifiers::NONE) => {
                        self.navigate_to_next_search_result()?;
                        return Ok(true);
                    }
                    (KeyCode::Char('N'), KeyModifiers::SHIFT) => {
                        self.navigate_to_previous_search_result()?;
                        return Ok(true);
                    }
                    _ => {}
                }
            }
        }

        // Existing hunk navigation logic
        match (key.code, key.modifiers) {
            (KeyCode::Char('n'), KeyModifiers::NONE) => {
                if self.copy_mode.is_some() {
                    // Don't conflict with copy mode
                    return Ok(false);
                }
                self.navigate_to_next_change()?;
                Ok(true)
            }
            (KeyCode::Char('N'), KeyModifiers::SHIFT) => {
                self.navigate_to_previous_change()?;
                Ok(true)
            }
            _ => Ok(false),
        }
    }

    pub fn handle_search_input_keys(&mut self, key: KeyEvent) -> Result<bool> {
        if let Some(ref mut search_state) = self.diff_search_state {
            if !search_state.is_input_mode {
                return Ok(false);
            }

            match (key.code, key.modifiers) {
                (KeyCode::Char(c), KeyModifiers::NONE) => {
                    search_state.query.push(c);
                    self.update_search_results()?;
                    Ok(true)
                }
                (KeyCode::Backspace, KeyModifiers::NONE) => {
                    search_state.query.pop();
                    self.update_search_results()?;
                    Ok(true)
                }
                (KeyCode::Enter, KeyModifiers::NONE) => {
                    search_state.is_input_mode = false;
                    if !search_state.results.is_empty() {
                        search_state.current_result = Some(0);
                        self.scroll_to_search_result(0)?;
                    }
                    Ok(true)
                }
                (KeyCode::Esc, _) => {
                    self.clear_diff_search();
                    Ok(true)
                }
                _ => Ok(false),
            }
        } else {
            Ok(false)
        }
    }
}
