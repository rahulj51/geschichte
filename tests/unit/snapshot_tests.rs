use insta::assert_debug_snapshot;
use std::path::Path;

mod syntax_highlighting_snapshots {
    use super::*;
    use geschichte::diff::syntax::highlight_line;
    
    #[test]
    fn test_rust_syntax_highlighting() {
        let rust_code = "fn main() { println!(\"Hello, world!\"); }";
        let file_path = Path::new("main.rs");
        
        let highlighted = highlight_line(rust_code, file_path);
        
        // Convert the highlighted spans to a more snapshot-friendly format
        let snapshot_data: Vec<(String, String)> = highlighted
            .into_iter()
            .map(|span| {
                let content = span.content.to_string();
                let style = format!("{:?}", span.style);
                (content, style)
            })
            .collect();
        
        assert_debug_snapshot!(snapshot_data);
    }
    
    #[test]
    fn test_json_syntax_highlighting() {
        let json_code = r#"{"name": "test", "value": 42, "enabled": true}"#;
        let file_path = Path::new("config.json");
        
        let highlighted = highlight_line(json_code, file_path);
        
        // Convert to snapshot-friendly format
        let snapshot_data: Vec<(String, String)> = highlighted
            .into_iter()
            .map(|span| {
                let content = span.content.to_string();
                let style = format!("{:?}", span.style);
                (content, style)
            })
            .collect();
        
        assert_debug_snapshot!(snapshot_data);
    }
}

mod app_state_snapshots {
    use super::*;
    use crate::common::helpers::create_test_app;
    use geschichte::app::{App, AppMode, FocusedPanel};
    use geschichte::cli::LayoutMode;
    use std::path::PathBuf;
    
    // Helper to create a snapshot-friendly version of App state
    #[derive(Debug)]
    #[allow(dead_code)] // Fields are used by Debug derive but linter doesn't see it
    struct AppSnapshot {
        repo_root: PathBuf,
        should_quit: bool,
        context_lines: u32,
        follow_renames: bool,
        first_parent: bool,
        mode: String, // Simplified representation
        commits_count: usize,
        selected_index: usize,
        layout_mode: LayoutMode,
        loading: bool,
        error_message: Option<String>,
        focused_panel: Option<FocusedPanel>,
        terminal_width: u16,
        terminal_height: u16,
    }
    
    impl From<&App> for AppSnapshot {
        fn from(app: &App) -> Self {
            let mode = match &app.mode {
                AppMode::FilePicker { context, .. } => {
                    format!("FilePicker({:?})", context)
                }
                AppMode::History { file_path, focused_panel } => {
                    format!("History(file: {:?}, focus: {:?})", file_path.file_name(), focused_panel)
                }
            };
            
            // Normalize the repo_root path to avoid snapshot differences due to temp directories
            let path_str = app.repo_root.to_string_lossy();
            let normalized_repo_root = if path_str.contains("/tmp") || path_str.contains("/var/folders") {
                PathBuf::from("/tmp/test_repo")
            } else {
                app.repo_root.clone()
            };
            
            AppSnapshot {
                repo_root: normalized_repo_root,
                should_quit: app.should_quit,
                context_lines: app.context_lines,
                follow_renames: app.follow_renames,
                first_parent: app.first_parent,
                mode,
                commits_count: app.commits.len(),
                selected_index: app.selected_index,
                layout_mode: app.layout_mode,
                loading: app.loading,
                error_message: app.error_message.clone(),
                focused_panel: app.get_focused_panel(),
                terminal_width: app.ui_state.terminal_width,
                terminal_height: app.ui_state.terminal_height,
            }
        }
    }
    
    #[test]
    fn test_initial_app_state() {
        let app = create_test_app();
        let snapshot = AppSnapshot::from(&app);
        
        assert_debug_snapshot!(snapshot);
    }
    
    #[test]
    fn test_app_state_after_resize() {
        let mut app = create_test_app();
        app.handle_resize(120, 40);
        
        let snapshot = AppSnapshot::from(&app);
        
        assert_debug_snapshot!(snapshot);
    }
}