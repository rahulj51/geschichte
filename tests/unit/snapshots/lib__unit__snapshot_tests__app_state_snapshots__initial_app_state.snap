---
source: tests/unit/snapshot_tests.rs
expression: snapshot
---
AppSnapshot {
    repo_root: "/tmp/test_repo",
    should_quit: false,
    context_lines: 3,
    follow_renames: false,
    first_parent: false,
    mode: "History(file: Some(\"test.txt\"), focus: Commits)",
    commits_count: 0,
    selected_index: 0,
    layout_mode: Auto,
    loading: false,
    error_message: None,
    focused_panel: Some(
        Commits,
    ),
    terminal_width: 80,
    terminal_height: 24,
}
